tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/IbmMq/IbmMqHeadersAdapter.cs

```diff
@@ -0,0 +1,86 @@
+// <copyright file="IbmMqHeadersAdapter.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Datadog.Trace.Headers;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.IbmMq;
+
+internal readonly struct IbmMqHeadersAdapter : IHeadersCollection
+{
+    private static readonly string[] EmptyValue = { };
+    private readonly IMqMessage _message;
+
+    public IbmMqHeadersAdapter(IMqMessage message)
+    {
+        _message = message;
+    }
+
+    /// <summary>
+    /// Used to normalize the property name, since IBM MQ only allows names which are valid Java identifiers
+    /// </summary>
+    /// <param name="name">Originally provided value</param>
+    /// <returns>Normalized name</returns>
+    private static string NormalizeName(string name)
+    {
+        var sb = new StringBuilder();
+        foreach (var c in name)
+        {
+            sb.Append(c is >= 'a' and <= 'z' or >= '0' and <= '9' ? c : '_');
+        }
+
+        var result = sb.ToString();
+        return result;
+    }
+
+    public IEnumerable<string> GetValues(string name)
+    {
+        try
+        {
+            // there's no way to check if the value exists,
+            // and reading non-existent value causes an exception
+            return new[] { _message.GetStringProperty(NormalizeName(name)) };
+        }
+        catch
+        {
+            return EmptyValue;
+        }
+    }
+
+    public void Set(string name, string value)
+    {
+        var normalizedName = NormalizeName(name);
+        RemoveNormalized(normalizedName);
+        _message.SetStringProperty(normalizedName, value);
+    }
+
+    public void Add(string name, string value)
+    {
+        var normalizedName = NormalizeName(name);
+        RemoveNormalized(normalizedName);
+        _message.SetStringProperty(normalizedName, value);
```

Comment:
I assume we can't just call `SetStringProperty` without the delete, and have it overwrite the value? 🙁

Reply:
SetStringProperty will not override the value, unfortunately :( You have to remove and re-add